### PR TITLE
feat: add keepAliveTTL to evict connection even if active

### PR DESCRIPTION
**Pull request by srijan02420**
Signed-off-by: Srijan Gupta <srijan02420@gmail.com>

Motivation:

If you are making http request to a remote-service with `keep-alive: true` the connection will never recycle if the request rate is high. In this scenario if a deployment took place for the remote service with blue-green configuration. Our client will still hit requests to old stack.
To solve this keeping a global TTL for active connections makes sense.
Similar to node's: https://www.npmjs.com/package/agentkeepalive#new-agentoptions

### CODE REVIEW

**Comment by vietj** [comment]
@srijan02420 there is no test for this feature

**Comment by vietj** [comment]
my point is that there is a new test for the connection pool, but no code is changed in connection pool (so at best it is a new test), the feature is added for HTTP/1.1 protocol and there is no test for with the HTTP client using HTTP/1

**Comment by srijan02420** [comment]
> @srijan02420 there is no test for this feature

@vietj I will add a tests for this in `testClientOptions`. Suggest where else I need to add tests.

**Comment by srijan02420** [comment]
@vietj I have added `testActiveConnectionTTL` with HttpClient using this option. Let me know if this works.

**Comment by srijan02420** [comment]
@vietj any update on this?

**Comment by vietj** [comment]
not yet, I need to understand the issue and see if that's the right thing to do

**Comment by srijan02420** [comment]
> not yet, I need to understand the issue and see if that's the right thing to do

Okay. Let me know if you need more info about the use case.

**Comment by vietj** [comment]
a review / analysis is set for 3.9.1

**Comment by vietj** [comment]
I have had a look at it and it seems an interesting feature, however while it seems to work in this case the current implementation is not correct because it will check the validity after the connection has been used which means that it needs to be used before closing the connection.

I believe this should be a feature of the pool itself and not the layer using the pool, so the pool would close such connection like it does today for expired connections.

**Comment by vietj** [comment]
Actually I believe it is correctly done :-), sorry for the incorrect analysis. Indeed the pool will take care of this. Allow me to do further analysis.

File: src/main/java/io/vertx/core/http/impl/Http1xClientConnection.java [hunk 1]
```diff
@@ -87,6 +87,7 @@
   private int keepAliveTimeout;
   private int seq = 1;
   private long bytesRead;
+  private long initialTimestamp = System.currentTimeMillis();
```
Comment (vietj) [inline src/main/java/io/vertx/core/http/impl/Http1xClientConnection.java:90]: we should call this only when the option is active, now need to call it otherwise

File: src/main/java/io/vertx/core/http/impl/Http1xClientConnection.java [hunk 2]
```diff
@@ -898,7 +899,13 @@ public void createStream(ContextInternal context, HttpClientRequestImpl req, Pro
   }
 
   private void recycle() {
-    long expiration = keepAliveTimeout == 0 ? 0L : System.currentTimeMillis() + keepAliveTimeout * 1000;
+    long now = System.currentTimeMillis();
```
Comment (vietj) [inline src/main/java/io/vertx/core/http/impl/Http1xClientConnection.java:902]: we should compute this only when the option is active

File: src/test/java/io/vertx/core/net/ConnectionPoolTest.java [hunk 3]
```diff
@@ -487,6 +487,26 @@ public void testDiscardExpiredConnections() {
     // assertWaitUntil(mgr::closed);
   }
 
+  @Test
```
Comment (vietj) [inline src/test/java/io/vertx/core/net/ConnectionPoolTest.java:490]: why is this test necessary ? is it a new test to ensure a behavior that was not tested previously ?

File: src/test/java/io/vertx/core/net/ConnectionPoolTest.java [hunk 4]
```diff
@@ -923,6 +944,14 @@ synchronized long recycle(long timestamp) {
       return inflight;
     }
 
+    synchronized long recycle(long keepAliveTimeout, long activeConnectionTTL, long currentTimestamp) {
```
Comment (vietj) [inline src/test/java/io/vertx/core/net/ConnectionPoolTest.java:947]: why is this necessary ?

File: src/main/java/io/vertx/core/http/HttpClientOptions.java [hunk 5]
```diff
@@ -196,6 +206,8 @@
   private int maxPoolSize;
   private boolean keepAlive;
   private int keepAliveTimeout;
+  private boolean isActiveConnectionTTL;
```
Comment (vietj) [inline src/main/java/io/vertx/core/http/HttpClientOptions.java:209]: I think those should be named differently, i.e in relation with the keep-alive feature, I think that keepAliveTTL is a better choice. Also these should be a single value that is either which means disabled or a positive value that is the TTL value.

Comment (srijan02420) [reply src/main/java/io/vertx/core/http/HttpClientOptions.java:209]: @vietj I will change `isActiveConnectionTTL` to `keepAliveTTL`.
I kept it as single variable initially and set default value as -1 because int in java cannot be null.
It felt it is sort of confusing.
Please suggest if I should revert back to that or you have something else in mind.

File: src/test/java/io/vertx/core/net/ConnectionPoolTest.java (hunk 3 above)
Comment (srijan02420) [reply src/test/java/io/vertx/core/net/ConnectionPoolTest.java:490]: It's actually not necessary. I got confused before. I am removing it.

File: src/test/java/io/vertx/core/net/ConnectionPoolTest.java (hunk 4 above)
Comment (srijan02420) [reply src/test/java/io/vertx/core/net/ConnectionPoolTest.java:947]: Yes not necessary. Removing it.

File: src/main/java/io/vertx/core/http/HttpClientOptions.java [hunk 6]
```diff
@@ -680,6 +688,35 @@ public HttpClientOptions setKeepAliveTimeout(int keepAliveTimeout) {
     return this;
   }
 
+  /**
+   * Is active connection ttl enabled on the client?
+   *
+   * @return {@code true} if enabled
+   */
+  public boolean isKeepAliveTTLEnabled() {
```
Comment (vietj) [inline src/main/java/io/vertx/core/http/HttpClientOptions.java:696]: can you annotate with `@GenIgnore` to not confuse the code generator ?

File: src/main/java/io/vertx/core/http/impl/Http1xClientConnection.java [hunk 7]
```diff
@@ -87,6 +87,7 @@
   private int keepAliveTimeout;
   private int seq = 1;
   private long bytesRead;
+  private long initialTimestamp;
```
Comment (vietj) [inline src/main/java/io/vertx/core/http/impl/Http1xClientConnection.java:90]: make it final

Comment (srijan02420) [reply src/main/java/io/vertx/core/http/impl/Http1xClientConnection.java:90]: @vietj if I will make it final then will have to initialise it.
Then will not be able to do 
```
if (options.isKeepAliveTTLEnabled())
      initialTimestamp = System.currentTimeMillis();
```

Comment (vietj) [reply src/main/java/io/vertx/core/http/impl/Http1xClientConnection.java:90]: you can init it with `0L`

Comment (srijan02420) [reply src/main/java/io/vertx/core/http/impl/Http1xClientConnection.java:90]: Yes but then cannot reassign it in constructor when options.isKeepAliveTTLEnabled() is true as 
```
initialTimestamp = System.currentTimeMillis();
```

**Comment by srijan02420** [comment]
@vietj done with the changes.
Any reason why `src/main/asciidoc/dataobjects.html` is not in .gitignore.

**Comment by vietj** [comment]
@srijan02420 I believe the work here is good but incomplete because validity cannot be checked when the connection is in the pool. I will improve the current pool so that now `recycle` is just a simple operation with no expiration timeout and the pool will make the actual validity check.

**Comment by vietj** [comment]
@srijan02420 can you try redo your contribution onto https://github.com/eclipse-vertx/vert.x/tree/pool-connection-validity-check-rework to check it works ?

**Comment by srijan02420** [comment]
@vietj sure. Doing that. Thanks.

**Comment by srijan02420** [comment]
@vietj check now.

**Comment by vietj** [comment]
I will finish this feature in master and then backport to 3.9 then you can update your PR for 3.9 and create a new one for master

File: src/main/java/io/vertx/core/http/impl/Http1xClientConnection.java [hunk 8]
```diff
@@ -897,8 +900,19 @@ public void createStream(ContextInternal context, HttpClientRequestImpl req, Pro
...
 
+  @Override
+  public boolean isValid() {
+    return expirationTimestamp > 0 && System.currentTimeMillis() <= expirationTimestamp;
+  }
+
   private void recycle() {
-    long expiration = keepAliveTimeout == 0 ? 0L : System.currentTimeMillis() + keepAliveTimeout * 1000;
-    listener.onRecycle(expiration);
+    if(!options.isKeepAliveTTLEnabled())
```
Comment (vietj) [inline src/main/java/io/vertx/core/http/impl/Http1xClientConnection.java:909]: this should actually be moved to `isValid` and when initialTimestamp > keepAliveTTL then return false

**Comment by vietj** [comment]
we also need a new test that check that the connection is closed from the pool when the connection is not active, so doing something like:

1/ make a client request when response comes, get the `HttpConnection` and set a `closeHandler` on the connection
2/ wait until the pool close the connection itself due to keepAliveTTL without an interaction with the pool

**Comment by srijan02420** [comment]
> we also need a new test that check that the connection is closed from the pool when the connection is not active, so doing something like:
> 
> 1/ make a client request when response comes, get the `HttpConnection` and set a `closeHandler` on the connection
> 2/ wait until the pool close the connection itself due to keepAliveTTL without an interaction with the pool

@vietj isn't `testKeepAliveTTL()` is doing this only?

**Comment by vietj** [comment]
I think there are two missing cases that are not covered:

1/ we need to test connections are evicted when keepAliveTimeout = 0 , in this case the vertx timer to evict connection from the pool is not started, so no check will be done

2/ HTTP/2 protocol is not implemented nor tested

**Comment by srijan02420** [comment]
You are right. I will write test for 1 and implement for H2 as well. Initially I was thinking because H2 uses only 1 connection and multiplex on that so there is nothing like pool. But I think it will make sense for that as well because it will help for blue/green deployments when H2 as wll.

**Comment by vietj** [comment]
It is still pooled actually (you pool a single connection) and handled in the same pool, the difference is that the concurrent for HTTP/2 is not the same.

**Comment by srijan02420** [comment]
@vietj I added test for `keepAliveTimeout = 0`. Please check if this was something you had in mind, I am still little bit confused if I understood that.
Meanwhile I am working for H2 implementation and test.

**Comment by vietj** [comment]
@srijan02420 actually I said something wrong, I thought that having keep alive timeout = 0 , disables the timer (see `HttpClientImpl` constructor) and it is actually the cleaner period.

I need to have a look again whether the timer is necessary when keep alive timeout = 0. The reason is that if timeout = 0 (which is never the case) I'm wondering if it's needed to close idle connections.

**Comment by srijan02420** [comment]
> @srijan02420 actually I said something wrong, I thought that having keep alive timeout = 0 , disables the timer (see `HttpClientImpl` constructor) and it is actually the cleaner period.
> 
> I need to have a look again whether the timer is necessary when keep alive timeout = 0. The reason is that if timeout = 0 (which is never the case) I'm wondering if it's needed to close idle connections.

Okay. Let me know.

**Comment by vietj** [comment]
@srijan02420 branch has been updated with new commits, I'll backport that to 3.9 branch

**Comment by srijan02420** [comment]
@vietj once you do that I will rebase with 3.9 branch and update this PR to that branch. And raise a separate PR for master.

**Comment by vietj** [comment]
@srijan02420 3.9 has been updated

**Comment by srijan02420** [comment]
@vietj rebased and updated.
If you want to add this to 3.9 as well. Will cheery-pick and create PR for that separately.

File: src/main/java/io/vertx/core/http/HttpClientOptions.java [hunk 9]
```diff
@@ -76,6 +77,11 @@
    */
   public static final int DEFAULT_KEEP_ALIVE_TIMEOUT = 60;
 
+  /**
+   *  Default keep alive time to live. -1 represents that it is disabled. Connection will never drop if active.
```
Comment (vietj) [inline src/main/java/io/vertx/core/http/HttpClientOptions.java:81]: _{@code -1}_

File: src/main/java/io/vertx/core/http/impl/Http1xClientConnection.java [hunk 10]
```diff
@@ -900,7 +902,8 @@ public void createStream(ContextInternal context, HttpClientRequestImpl req, Pro
 
   @Override
   public boolean isValid() {
-    return expirationTimestamp == 0 || System.currentTimeMillis() <= expirationTimestamp;
+    long now = System.currentTimeMillis();
+    return expirationTimestamp == 0 || now <= expirationTimestamp && (!options.isKeepAliveTTLEnabled() || options.getKeepAliveTTL() * 1000 > now - initialTimestamp);
```
Comment (vietj) [inline src/main/java/io/vertx/core/http/impl/Http1xClientConnection.java:906]: can you surround with parenthesis the two expressions right and left of `&&` to make it more clear ?

File: src/test/java/io/vertx/core/http/HttpTest.java [hunk 11]
```diff
@@ -5048,6 +5048,19 @@ public void testKeepAliveTimeout() throws Exception {
     testKeepAliveTimeout(options, 1);
   }
 
+  @Test
+  public void testKeepAliveTTL() throws Exception {
+    server.requestHandler(req -> {
+      req.response().end();
+    });
+    HttpClientOptions options = createBaseClientOptions()
```
Comment (vietj) [inline src/test/java/io/vertx/core/http/HttpTest.java:5056]: I believe we should use longer keepAliveTimeout values here like 3 seconds

Comment (srijan02420) [reply src/test/java/io/vertx/core/http/HttpTest.java:5056]: Sorry didn't understand you here.
I used 300 sec as keepAliveTimeout and 3 sec as keepAliveTTL. So connection should close after 3 sec.

**Comment by vietj** [comment]
out of my mind @srijan02420 , what is a good TTL value for blue-green deployments when using this feature ?

**Comment by srijan02420** [comment]
> out of my mind @srijan02420 , what is a good TTL value for blue-green deployments when using this feature ?
@vietj 5 mins i.e. 300 sec seems to be a fine number. Now I understood you want to keep that value in the test. They only problem is the test will have to wait that long.

**Comment by vietj** [comment]
@srijan02420 a small value is fine for tests, I was asking in case somebody asks me what is a good value to use in practice :-)

**Comment by srijan02420** [comment]
@vietj should we merge it. My team desperately need this feature. :)

**Comment by vietj** [comment]
@srijan02420 sure it will be merged, don't worry

File: src/test/java/io/vertx/core/http/HttpTest.java (hunk 11 above)
Comment (vietj) [reply src/test/java/io/vertx/core/http/HttpTest.java:5056]: ok sorry

**Comment by vietj** [comment]
@srijan02420 I am actually not sure the current approach we chose for this feature is correct. I think we should try this test:

1/ define a pool with a max connection size = 1 and keepAliveTTL = 1 (1 second)
2/ make two HTTP requests
3/ the HTTP server will wait for more than one second (1.5 second)
4/ the pool should close the first connection before serving the 2nd request still waiting in the pool

I am not sure this is currently satisfied, can you add a test for this case ?

**Comment by srijan02420** [comment]
> @srijan02420 I am actually not sure the current approach we chose for this feature is correct. I think we should try this test:
> 
> 1/ define a pool with a max connection size = 1 and keepAliveTTL = 1 (1 second)
> 2/ make two HTTP requests
> 3/ the HTTP server will wait for more than one second (1.5 second)
> 4/ the pool should close the first connection before serving the 2nd request still waiting in the pool
> 
> I am not sure this is currently satisfied, can you add a test for this case ?

Sure. Adding.

**Comment by vietj** [comment]
thanks

File: src/test/java/io/vertx/core/http/HttpTest.java [hunk 12]
```diff
@@ -5061,6 +5061,49 @@ public void testKeepAliveTTL() throws Exception {
     testKeepAliveTimeout(options, 1);
   }
 
+  @Test
+  public void testKeepAliveTTLWith2Req() throws Exception {
```
Comment (srijan02420) [inline src/test/java/io/vertx/core/http/HttpTest.java:5065]: @vietj this is something you had in mind right?

Comment (vietj) [reply src/test/java/io/vertx/core/http/HttpTest.java:5065]: the test needs to check that each connection obtained for the two responses are different instances (i.e not the same physical connections)

File: src/main/java/io/vertx/core/net/impl/clientconnection/Pool.java [hunk 13]
```diff
@@ -310,6 +291,26 @@ private Runnable nextTask() {
         };
       }
     }
+    if (waitersQueue.size() > 0) {
```
Comment (srijan02420) [inline src/main/java/io/vertx/core/net/impl/clientconnection/Pool.java:294]: @vietj we will have to check if connections are valid and expire them before assigning the connection.

File: src/test/java/io/vertx/core/http/HttpTest.java (hunk 12 above)
Comment (srijan02420) [reply src/test/java/io/vertx/core/http/HttpTest.java:5065]: Added test for same.

**Comment by vietj** [comment]
@srijan02420 I see your suggestion about validity check I think it is valid but only in the case of an HTTP/1 keep alive connection. It will not work for HTTP/2 and the connection would remain in the pool.

So it seems for this case we would need to do it differently using a shutdown phase, i.e the connection is still in the pool but will not accept any new connection until it is closed.

**Comment by vietj** [comment]
The other reason I think it should use shutdown is that connection max life time can be achieved anytime and has no relationship with the pool itself, whereas keep alive timeout is actually has an intrinsic relationship with the pool.

So I believe that the way to go would be to implement `HttpConnection#shutdown()` for HTTP/1 (it is for HTTP/2) to have a proper eviction of the connection from the pool.

This way the connection max life time can be achieved simply by setting a timer and call shutdown when the timer fires.

**Comment by srijan02420** [comment]
> The other reason I think it should use shutdown is that connection max life time can be achieved anytime and has no relationship with the pool itself, whereas keep alive timeout is actually has an intrinsic relationship with the pool.
> 
> So I believe that the way to go would be to implement `HttpConnection#shutdown()` for HTTP/1 (it is for HTTP/2) to have a proper eviction of the connection from the pool.
> 
> This way the connection max life time can be achieved simply by setting a timer and call shutdown when the timer fires.

Understood. I will try and do it.

**Comment by srijan02420** [comment]
@vietj whenever I am doing something like
```
vertx.setTimer(client.getOptions().getKeepAliveTTL(), a -> {
      if (handler.connection().numActiveStreams() ==  0)
                shutdown(0L);
}
```
I am getting `io.vertx.core.VertxException: Connection was closed` ie I am trying to shutdown connection before `responseEnded = true;`. But I am not able to figure where should I call graceful shutdown.

**Comment by vietj** [comment]
@srijan02420 the shutdown needs to be implemented for HTTP/1

**Comment by vietj** [comment]
I will have a look at it

**Comment by srijan02420** [comment]
> @srijan02420 the shutdown needs to be implemented for HTTP/1

Yes above was for H2 only. Shutdown some is not waiting for response to write for my test apparently.
Its not shutting down gracefully.

**Comment by srijan02420** [comment]
@vietj I think connection max life time is also related to pool as keepAliveTmeout is because as pool understand that this connection sat idle for long and now should be evicted. Pool should understand the connection hit its life time limit and should be eveicted.
Other way to look at it is that option keep-alive can be used with pool only so keepAliveTTL should be used with pool only. Without keep-alive true keepAliveTTL doesn't make sense.

**Comment by vietj** [comment]
@srijan02420 for now I am quite cautious because it seemed like an easy feature to implement but in fact it is not and I don't want to bring a new feature with bugs. For now I will push a branch that implements shutdown for HTTP/1.1 connection and you can use it for your case by using a timer to shutdown the connection after a while (it is trivial).

**Comment by vietj** [comment]
@srijan02420 can you have a look at https://github.com/eclipse-vertx/vert.x/tree/http-1-shutdown

**Comment by srijan02420** [comment]
> @srijan02420 can you have a look at https://github.com/eclipse-vertx/vert.x/tree/http-1-shutdown

@vietj I checked.

1/ Using this should I do something like
```
vertx.setTimer(client.getOptions().getKeepAliveTTL(), a -> {
      shutdown(0L);
}
```
inside `Http1xClientConnection` constructor?

2/ History of `http-1-shutdown` is different from history of `master`. Showing a lot of conflicts while merging.

**Comment by vietj** [comment]
@srijan02420 for now we would not implement this feature until shutdown becomes more robust, so in your application you should manage yourself the timers and the shutdown calls

**Comment by srijan02420** [comment]
Oh okay. Let me know if I can help in improving shutdown.

**Comment by vietj** [comment]
you can test the feature extensively and report feedback that will be useful

**Comment by srijan02420** [comment]
with shutdown you mean?
Will start using once its merged.

**Comment by vietj** [comment]
it needs some work on documentation and also port to master branch

**Comment by vietj** [comment]
The branch has been merged @srijan02420 https://github.com/eclipse-vertx/vert.x/pull/3384

**Comment by srijan02420** [comment]
@vietj thanks. I will start using it and share results. Let me know whenever you think we can start re-work for this feature.

**Comment by vietj** [comment]
for now I will close this PR to clean up things

**Comment by vietj** [comment]
thanks for your time @srijan02420 and please give feedback on the shutdown once you have tested it
